serial/pet/driver.s

```diff
@@ -0,0 +1,730 @@
+                sta VIA_PCR
+                ; Set VIA interrupts so that our timer is the only interrupt source
+                lda #$7F              ; Clear all interrupt flags
+                sta VIA_IER
+                lda #$C2              ; Enable Timer 1 interrupt and CA1 interrupt
+                sta VIA_IER
+
+                lda #0
+                sta VIA_T1CL
+                sta VIA_T1CH          ; Need to clear high before writing latch
+                                      ; Otherwise it seems to fail half the time?
+
+                ldx #BAUD             ; Set up timers based on BAUD
+                lda BaudTblLo,X       ; Set interrupt timer
+                sta VIA_T1LL
+                lda BaudTblHi,X
+                sta VIA_T1LH
+
+                lda KbdPollIntTbl,X   ; Set keyboard polling interval based
+                sta KbdPollIntrvl     ; on current baud/timer rate
+                sta KbdPollCnt
+
+                cli
+                rts
+
+;-----------------------------------------------------------------------------------
+; GetSerialChar: Will fetch a character from the receive buffer and store it into A.
+; If no data is available, SER_ERR_NO_DATA is returned in X/Y.
+;-----------------------------------------------------------------------------------
+
+GetSerialChar:
+                ldx RxBufReadPtr
+                cpx RxBufWritePtr
+                beq @nodata           ; No character available
+                lda RX_BUF,X          ; New character
+                inc RxBufReadPtr      ; Acknowledge byte by incrementing
+                rts
+
+@nodata:
+                lda #$ff
+                ldx #<SER_ERR_NO_DATA
+                ldy #>SER_ERR_NO_DATA
+                rts
+
+;-----------------------------------------------------------------------------------
+; PutSerialChar: Output character in A. This blocks if we're still sending another
+; char.
+;-----------------------------------------------------------------------------------
+
+PutSerialChar:
+                ldx TxNewFlag
+                bne PutSerialChar     ; Loop till we can send a character
+                sta TxNextByte
+                lda #$FF
+                sta TxNewFlag
+                rts
+
+
+;-----------------------------------------------------------------------------------
+; StartSerial: Start serial communication. OpenSerial must have been called already.
+;-----------------------------------------------------------------------------------
+
+StartSerial:
+                lda #0
+                sta RxBufWritePtr
+                sta RxBufReadPtr
+
+                rts
+
+;-----------------------------------------------------------------------------------
+; CloseSerial: Teardown serial comms. We try to restore everything to its original
+; state.
+;-----------------------------------------------------------------------------------
+
+CloseSerial:    
```

Comment:
A few comments on what all these masks and values are doing as they are stored in the PIA/VIA might be prudent?

Reply:
I've copied this code verbatim from the PETTERM tear down code, so I could pass that comment on to the authors of PETTERM. :) I assume these values are all retrieved from the PET KERNAL initialization code, either by copying parts of the KERNAL ROM or establishing what the respective values are immediately after the machine starts up.